<p align="center">
<img src='../../img/VerticaMLPython.png' width="180px">
</p>

# Vertica ML Python Exercise 2

During this exercice, we will see:
<ul>
    <li> How to detect outliers using 4 different techniques (DBSCAN, KMeans, LOF and Normal Distribution)
    <li> How to fill outliers
    <li> How to drop outliers
    <li> Find the KMeans K
    <li> How to realize different scatter plots
</ul>
## Initialization

Let's create a cursor using the vertica_cursor function

In [1]:
from vertica_ml_python.utilities import vertica_cursor
cur = vertica_cursor("VerticaDSN")

During this study, we will work with the heart disease dataset. You'll need to read the 'heart.csv' file.

In [7]:
from vertica_ml_python.utilities import read_csv
from vertica_ml_python.utilities import drop_table
drop_table('heart', cur)
heart = read_csv('heart.csv', cur)
print(heart)

The table heart was successfully dropped.
The table public.heart has been successfully created.


<object>  Name: heart, Number of rows: 270, Number of columns: 14


This dataset contains many information of 270 patients including:
<ul>
    <li><b>age:</b> age in years</li>
    <li><b>thalach:</b> maximum heart rate achieved</li>
    <li><b>trestbps:</b> resting blood pressure (in mm Hg on admission to the hospital) </li>
    <li><b>fbs:</b> (fasting blood sugar > 120 mg/dl) (1 = true; 0 = false) </li>
    <li><b>slope:</b> the slope of the peak exercise ST segment (Value 1: upsloping; Value 2: flat; Value 3: downsloping) </li>
    <li><b>ca:</b> number of major vessels (0-3) colored by flourosopy </li>
    <li><b>num:</b> diagnosis of heart disease (angiographic disease status) (Value 1: < 50% diameter narrowing; Value 2: > 50% diameter narrowing)</li>
    <li><b>sex:</b> sex (1 = male; 0 = female) </li>
    <li><b>cp:</b> hest pain type (Value 1: typical angina; Value 2: atypical angina; Value 3: non-anginal pain; Value 4: asymptomatic) </li>
    <li><b>exang:</b> exercise induced angina (1 = yes; 0 = no) </li>
    <li><b>restecg:</b> resting electrocardiographic results (Value 0: normal; Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV); Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria) </li>
    <li><b>chol:</b> serum cholestoral in mg/dl </li>
    <li><b>oldpeak:</b> ST depression induced by exercise relative to rest </li>
    <li><b>thal:</b> 3 = normal; 6 = fixed defect; 7 = reversable defect </li>
</ul>

The purpose is to find people having heart complications compare to the rest. 

## Outliers Detection

Let's explore the data by displaying descriptive statistics of all the columns.

In [9]:
heart.describe()

<object>

We will only work with the maximum heart rate achieved (thalach) and the serum cholestoral (chol). We will be able to identify outliers with different techniques and visualize the results.

<b>Question 1: </b>Draw the Scatter Plot of the two variables. What do you notice ?

<b>Question 2: </b>Add the Global Outliers of the distribution using the 'outliers' method. Draw again the Scatter Plot with this time marking the outliers. Find visually the number of outliers and explain the result. 

You can drop or fill these outliers by using respectively the 'drop_outliers' and 'fill_outliers' methods. It will only consider global outliers of the distribution. Otherwise you can use the 'filter' method to do it.

<b>Question 3: </b>Use the DBSCAN with epsilon = 20 and min_samples = 10. The outliers will be labeled with the class -1. What do you notice ?

<b>Question 4: </b>Now, use the Local Outlier Factor Algorithm and draw the LOF graph. Build a Virtual Dataframe using the created table and create a new label based on the LOF (>1.5 = outliers)

<b>Question 5: </b>The last technique to find outliers is using the KMeans algorithm. However, we never know the number of clusters in advance. Use the elbow curve to find it. Create a KMeans model with this number. Draw the model prediction and identify the outliers.

<b>Question 6: </b>Use your favourite technique to identify the outliers and draw histograms to see what part of the population is affected by huge complications.